arangod/Aql/HashedCollectExecutor.cpp

```diff
@@ -0,0 +1,299 @@
+    // NOTE: this might impact the performance
+    if (input.isInitialized()) {
+      _lastInitializedInputRow = input;
+
+      consumeInputRow(input);
+    }
+  }
+
+  // initialize group iterator for output
+  if (!_producingOutput) {
+    _producingOutput = true;
+    // setup the iterator
+    _currentGroup = _allGroups.begin();
+  }
+
+  return ExecutionState::DONE;
+}
+
+std::pair<ExecutionState, NoStats> HashedCollectExecutor::produceRow(OutputAqlItemRow& output) {
+  TRI_IF_FAILURE("HashedCollectExecutor::produceRow") {
+    THROW_ARANGO_EXCEPTION(TRI_ERROR_DEBUG);
+  }
+
+  if (!_isInitialized) {
+    // fetch & consume all input and initialize output cursor
+    ExecutionState state = init();
+    if (state == ExecutionState::WAITING) {
+      return {state, NoStats{}};
+    }
+    TRI_ASSERT(state == ExecutionState::DONE);
+    _isInitialized = true;
+  }
+
+  // produce output
+  if (_currentGroup != _allGroups.end()) {
+    writeCurrentGroupToOutput(output);
+    _currentGroup++;
+  }
+
+  ExecutionState state = _currentGroup != _allGroups.end() ? ExecutionState::HASMORE
+                                                           : ExecutionState::DONE;
+
+  return {state, NoStats{}};
+}
+
+// if no group exists for the current row yet, this builds a new group.
+std::pair<std::unique_ptr<HashedCollectExecutor::AggregateValuesType>, std::vector<AqlValue>>
+HashedCollectExecutor::buildNewGroup(InputAqlItemRow& input, size_t n) {
+  std::vector<AqlValue> group;
+  group.reserve(n);
+
+  // copy the group values before they get invalidated
+  for (size_t i = 0; i < n; ++i) {
+    group.emplace_back(input.stealValue(_infos.getGroupRegisters()[i].second));
+  }
+
+  auto aggregateValues = std::make_unique<AggregateValuesType>();
+  aggregateValues->reserve(_aggregatorFactories.size());
+
+  for (auto const& it : _aggregatorFactories) {
+    aggregateValues->emplace_back((*it)(_infos.getTransaction()));
+  }
+  return std::make_pair(std::move(aggregateValues), group);
+}
+
+// finds the group matching the current row, or emplaces it. in either case,
+// it returns an iterator to the group matching the current row in
+// _allGroups. additionally, .second is true iff a new group was emplaced.
+std::pair<decltype(HashedCollectExecutor::_allGroups)::iterator, bool>
+HashedCollectExecutor::findOrEmplaceGroup(InputAqlItemRow& input) {
+  std::vector<AqlValue> groupValues;  // TODO store groupValues locally
+  size_t const n = _infos.getGroupRegisters().size();
+  groupValues.reserve(n);
+
+  // for hashing simply re-use the aggregate registers, without cloning
+  // their contents
+  for (size_t i = 0; i < n; ++i) {
+    groupValues.emplace_back(input.getValue(_infos.getGroupRegisters()[i].second));
+  }
+
+  auto it = _allGroups.find(groupValues);
+
+  if (it != _allGroups.end()) {
+    // group already exists
+    return {it, false};
+  }
+
+  // must create new group
+  std::unique_ptr<AggregateValuesType> aggregateValues;
```

Comment:
Similarly, use `GroupValueType` instead of `std::unique_ptr<AggregateValuesType>`, please.